src/hepattn/models/decoder.py

```diff
@@ -68,42 +76,52 @@ def forward(self, x: dict[str, Tensor], input_names: list[str]) -> tuple[dict[st
         num_constituents = x["key_embed"].shape[-2]
         self.log_step += 1
 
-        outputs = {}
+        if (self.key_posenc is not None) or (self.query_posenc is not None):
+            x["query_posenc"], x["key_posenc"] = self.generate_positional_encodings(x)
+        if not self.preserve_posenc:
+            x["query_embed"], x["key_embed"] = self.add_positional_encodings(x)
+
+        outputs: dict[str, dict] = {}
 
         for layer_index, decoder_layer in enumerate(self.decoder_layers):
             outputs[f"layer_{layer_index}"] = {}
 
-            attn_masks = {}
-            query_mask = None
-
-            for task in self.tasks:
-                if not task.has_intermediate_loss:
-                    continue
-
-                # Get the outputs of the task given the current embeddings
-                task_outputs = task(x)
-
-                # Update x with task outputs for downstream use
-                if isinstance(task, IncidenceRegressionTask):
-                    x["incidence"] = task_outputs[task.outputs[0]].detach()
-                if isinstance(task, ObjectClassificationTask):
-                    x["class_probs"] = task_outputs[task.outputs[0]].detach()
+            if self.preserve_posenc:
+                x["query_embed"], x["key_embed"] = self.add_positional_encodings(x)
 
-                outputs[f"layer_{layer_index}"][task.name] = task_outputs
-
-                # Collect attention masks from tasks
-                task_attn_masks = task.attn_mask(task_outputs)
-                for input_name, attn_mask in task_attn_masks.items():
-                    if input_name in attn_masks:
-                        attn_masks[input_name] |= attn_mask
-                    else:
-                        attn_masks[input_name] = attn_mask
+            attn_masks: dict[str, torch.Tensor] = {}
+            query_mask = None
 
-                # Collect query masks
-                if self.use_query_masks:
-                    task_query_mask = task.query_mask(task_outputs)
-                    if task_query_mask is not None:
-                        query_mask = task_query_mask if query_mask is None else query_mask | task_query_mask
+            if self.tasks is not None:
```

Comment:
`self.tasks` should never be `None` at this stage, so I would remove this `if` to reduce indentation and have an explicit error on the next line if something has gone wrong